lib/grammers-client/src/client/auth.rs

```diff
@@ -261,6 +265,58 @@ impl Client {
             Err(InvocationError::Rpc(RpcError { name, .. })) if name.starts_with("PHONE_CODE_") => {
                 Err(SignInError::InvalidCode)
             }
+            Err(InvocationError::Rpc(RpcError { name, .. }))
+                if name.starts_with("SESSION_PASSWORD_NEEDED") =>
+            {
+                Err(SignInError::PasswordRequired)
+            }
+            Err(error) => Err(SignInError::Other(error)),
+        }
+    }
+
+    pub async fn two_factor_auth(
+        &mut self,
+        password: Vec<u8>,
+    ) -> Result<tl::types::User, SignInError> {
+        let request = tl::functions::account::GetPassword {};
+
+        let data: tl::enums::account::Password = match self.invoke(&request).await {
+            Ok(x) => x.into(),
+            Err(error) => return Err(SignInError::Other(error)),
+        };
+
+        let password_information: tl::types::account::Password = match data {
+            tl::enums::account::Password::Password(pass) => pass,
+        };
+
+        let tl::types::PasswordKdfAlgoSha256Sha256Pbkdf2Hmacsha512iter100000Sha256ModPow { salt1, salt2, g, p } = match password_information.current_algo.unwrap() {
+            tl::enums::PasswordKdfAlgo::Unknown => panic!("Unknown KDF (most likely, the client is outdated and does not support the specified KDF algorithm)"),
+            tl::enums::PasswordKdfAlgo::Sha256Sha256Pbkdf2Hmacsha512iter100000Sha256ModPow(alg) => alg,
+        };
+
+        let g_b = password_information.srp_b.unwrap();
+        let a: Vec<u8> = password_information.secure_random;
+
+        let (m1, g_a) = calculate_2fa(salt1, salt2, g, p, g_b, a, password);
+
+        let check_password = tl::functions::auth::CheckPassword {
+            password: tl::enums::InputCheckPasswordSrp::Srp(tl::types::InputCheckPasswordSrp {
+                srp_id: password_information.srp_id.unwrap(),
+                a: g_a,
+                m1,
+            }),
+        };
+
+        match self.invoke(&check_password).await {
+            Ok(tl::enums::auth::Authorization::Authorization(x)) => {
+                // Safe to unwrap, Telegram won't send `UserEmpty` here.
+                Ok(x.user.try_into().unwrap())
+            }
+            Ok(tl::enums::auth::Authorization::SignUpRequired(x)) => {
```

Comment:
I do not think Telegram can send a sign up required here. I think we should panic if this variant is returned (kind of "unreachable", as it's done in other places; but not an actual unreachable because we never know what Telegram might do).